Publish documentation

markrtuttle opened the pull request:
This pull request proposes a method of publishing documentation in the diffblue/cbmc repository to the diffblue.github.io/cbmc web site.  See [this comment](https://github.com/diffblue/cbmc/issues/7006#issuecomment-1245848220) for a discussion of the goals.  See [markrtuttle.github.io/cbmc-documentation](https://markrtuttle.github.io/cbmc-documentation) for a demonstration.

After merging this pull request, the diffblue/cbmc repository must be configured to publish pages from the branch `gh-pages` to diffblue.github.io/cbmc.  See [github instructions](https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site) to configure from within the account settings (basically two clicks).

There are many known issues.  A lot of these comments have to do with using pandoc to remove doxygen issues from markdown files before running doxygen.  It is nearly impossible to write markdown files that format cleanly with doxygen and two versions of pandoc and still read well on github as rendered markdown (and still work well with the javascript rendering www.cprover.org/cprover-manual).  The biggest issue is that doxygen is written to render markdown snippets from code comments, but does not handle well a collection of basic markdown files that interlink among themselves using classical markdown syntax. Most important, doxygen wants explicitly-declared, globally-unique labels in markdown (that is not supported by classical markdown) for links, and does not follow markdown syntax for linking to sections within a markdown file.  We need to make a decision about whether we are writing markdown or doxygen markdown.  I recommend (sigh) doxygen markdown and using `\page unique-label Heading` in place of `# Heading`.  For section labels, without requiring unique section names, the only option appears to be `\section unique-label Section` in place of `## Section`, but that is a massive revision (or just give up on linking directly to section headers).

The modern brew doxygen is 1.9.5 and the ubuntu-latest doxygen on github is 1.8.17.  The older version 1.8.17 appears to differ slightly between brew and ubuntu (eg, volume of progress information printed during execution).  The newer version 1.9.5 issues warnings that the older version 1.8.17 does not.  Differences include some unused variables in doxyfile that are now obsolete, and 500+ new warnings of the form

```
cbmc/src/util/std_types.h:736: 
   warning: unable to resolve reference to 'code_typet.\ifile' for \ref command
```

* This pull request restores two files that may have been deleted from develop:
    * doc/architectural/restrict-function-pointer.md 
    * doc/architectural/nondet-volatile.md

* goto-cc has broken links (what should they link to?):

  ```
  -   \ref man_goto-cc-linux "Linux Kernel"
  -   \ref man_goto-cc-apache "Apache HTTPD"
  ```

* test_suite.md a broken link (what should it link to?):

  ```
  [pid\_test\_suites.xml](pid_test_suites.xml)
  ```

* pandoc cannot handle lists or rules that are not preceded by a blank line. In particular, it will try to interpret text immediately following the horizontal rule `---` as yaml.

* pandoc --read markdown sometimes treats doxygen commands like \ingroup as latex commands.  Using pandoc --read markdown_phpextra avoids this problem.

* pandoc --read markdown_phpextra sometimes deletes or modifies fenced code blocks with syntax highlighting like ```c. Using pandoc --read markdown avoids this problem.

* doxygen cannot handle links `[hyper link](url)` that break over two lines. Using pandoc --wrap none solves this problem, but now what to use for --read and --write formats (see above).

* cprover-manual has mutliple sections with same title resulting in nonunique labels for doxygen.  Doxygen cannot link to internal sections with `[section](#section)`.  Doxygen requires globally unique section labels.  This is not supported in straight markdown.  The pandoc --write markdown_phpextra will generate the needed section labels, but doxygen complains because they are not unique.  For example, in the contracts pages, there are multiple sections named "Overview".

* cprover-manual does not have top-level sections, and doxygen complains it has subsections outside of sections, etc.

Commit:
Correct doxygen \page commands in architectural documentation

* Replace top-level section headings with \page command
* Ensure a space between \ingroup and \page for pandoc parsing
* Remove link to cprover-manual index from memory-analyzer,
  restrict-function pointer, etc, and make these pages subpages
  of the top-level architectural page

Commit:
Make cprover-manual format nicely with doxygen

* Format index.md as nested lists instead of a list of section headings
* Correct broken links
* Add blank link before all lists for pandoc parsing
* Some end-of-line whitespace removed

Commit:
Make three README files format with doxygen/pandoc

* Add README files of doc/API and src/goto-harness
* Add surround markdown horizontal rule --- with blank lines for
  pandoc parsing (pandoc will try to interpret text following ---
  as yaml).

Commit:
Make satabs user manual a first-class page

Commit:
Add markdown preprocessing scripts for use before doxygen formatting

* Append "last modified" dates to the end of all markdown files.
* Use a pandoc filter to rewrite links in cprover-manual (links
  originally intended to work with the cprover.org/cprover-manual
  javascripts).
* Add a script to repair fenced code blocks written by pandoc

Commit:
Add doxyfiles for miscellaneous sets of markdown files

* doc/ADR
* doc/API
* doc/assets
* doc/cprover-manual

Commit:
Add top-level documtation pages

Add top-level user guide, reference guide, developer guide pages for
the new published documentation.  The top-level doxyfile is the
original src/doxyfile with minor modification (eg, project title is
now "CBMC").

Commit:
Add cbmc documentation publication workflow

tautschnig commented on doc/architectural/front-page.md:
```diff
@@ -1,5 +1,5 @@
-CProver Developer Documentation
-=====================
+
```
Nit pick: blank line at the top of the file?

markrtuttle replied:
Done.

tautschnig commented on doc/architectural/nondet-volatile.md:
```diff
@@ -0,0 +1,136 @@
+\ingroup module_hidden 
```
Restoring this is fine, but I had intentionally removed it for its contents was added to the man page (of goto-instrument).

markrtuttle replied:
Removed (and links from user guide and architectural guide).

tautschnig commented on doc/architectural/restrict-function-pointer.md:
```diff
@@ -0,0 +1,182 @@
+
+\page restricting-function-pointers Restricting function pointers
```
As above: I'm ok with restoring it, but I do note that the man page has this content now.

markrtuttle replied:
Removed (and links from user guide and architectural guide).

tautschnig commented on doc/cprover-manual/index.md:
```diff
@@ -1,34 +1,34 @@
 
-## 1. [Introduction](introduction/)
+1. [Introduction](introduction/)
 
-## 2. [Installation](installation/)
+2. [Installation](installation/)
```
@kroening Will this render ok on cprover.org?

markrtuttle replied:
FWIW, I did mention this to @kroening  (replacing the sequence of headers with a list) and he did not object.

tautschnig commented on doc/doxygen-root/doxygen-markdown/markdown.sh:
```diff
@@ -0,0 +1,57 @@
+# Note: Need to read markdown as markdown_phpextra and not default
+# markdown to preserve doxygen pragmas like \ingroup.
+
+# Bug: This is currently interacting badly with \dot in cprover markdown
```
To what extent should we be concerned about this? Is there any plan to do something about this?

markrtuttle replied:
This preprocessing is required on the cprover-manual pages as long as they are served up by javascript on prover.org.  I did a quick scan of the architectural pages, and I saw no obvious problems caused by omitting this preprocessing on the current set of pages.  But I think doxygen's brittle treatment of markdown files is a potential issue, and I think we would be better off just deciding to write doxygen-friendly markdown and just living with subpar rendering by other tools (like on github).

tautschnig commented on doc/doxygen-root/doxyfile:
```diff
@@ -0,0 +1,2423 @@
+# Doxyfile 1.8.11
```
1) Why is this a different version than the other files?
2) Should the existing doxyfile be removed?

markrtuttle replied:
Updated doxyfile to version generated by doxygen 1.8.17 and removed src/doxyfile that was the original doxyfile and the basis for the new doxyfile.

markrtuttle replied:
Turns out deleting src/doxyfile causes the run_doxygen check to fail, so I've restored it until that check is disabled.

tautschnig replied:
Could you perhaps add a commit that removes that file and adjusts the CI job that uses `run_doxygen` to do something that is meaningful with the new documentation set-up? The primary goal of the CI job is to make sure people don't add documentation that would cause new Doxygen warnings to appear.

Commit:
Remove a blank line from the top of a markdown file

Commit:
Remove "nondet volatile" and "function pointer" pages from user guide

The content of nondet-volatile.md and restrict-function-pointer.md in
doc/architectural has already been moved to the goto-instrument man page.
Remove this also from architectural main page front-page.md

Commit:
Update doc/doxygen-root/doxyfile to version 1.8.17

This doxygen file is based on src/doxyfile, and src/doxyfile can be
removed once the run_doxygen continuous integration check is disabled.

jimgrundy commented:
I'd like to see this adopted so that documentation can be maintained more easily and directly from GitHub, and browsable within GitHub and local repos. Further, I'd like to see is declare "Doxygen" and "MarkDown as spoken by Doxygen" become our expected documentation method, which would allow us to strip away a little of the scripting complexity Mark added to support the Markdown as it is today. But, that can wait. Please let's get this in.

NlightNFotis commented:
@jimgrundy Hi Jim, this is on my plate for today - will get this reviewed, but then I'm not sure who else this needs reviews from. I'll see if it needs a couple more from our end if I can chase them further (but please be aware that we've been hit by unexpected illness in our team, so we're missing some of our usual bandwidth).

NlightNFotis commented on doc/API/README.md:
```diff
@@ -0,0 +1,3 @@
+# CProver APIs
```
This should be changed to something like:

```
\page apis CPROVER APIs
```

markrtuttle replied:
This particular proposal won't work because this is the main markdown file for `doc/API`.  The folders `doc/API`, `doc/ADR`, and `doc/assets` are built separately because they don't seem to be linked into the main CPROVER documentation.

I've taken the liberty of converting `doc/API` and `doc/ADR` to the doxygen page/subpage syntax.  

I was going to do the same for `doc/assets`, but there is an xml-spec.md that is not linked to by the top-level README, and I wasn't sure which of the various specs are considered authoritative.  

@NlightNFotis

NlightNFotis commented on doc/doxygen-root/README.md:
````diff
@@ -0,0 +1,60 @@
+* On MacOS,
+  we recommend installing with [brew](https://brew.sh/):
+  ```
+  brew install doxygen graphviz
````
Does this need `pandoc` as well?

I tried without and got some errors?

markrtuttle replied:
I've added pandoc to the instructions.  I omitted it because it is already installed on the GitHub ubuntu runner.  But good catch.

NlightNFotis commented on doc/doxygen-root/README.md:
````diff
@@ -0,0 +1,60 @@
+To build the documentation, run
+
+```
+doxygen
````
There's also a `Makefile`.

I tried running the `default` action of the `Makefile` first, but got a very broken version of the documentation (probably because I was missing stuff from the dependencies?) - which one is the defacto way a developer builds documentation locally? Just running `Doxygen`, or the `Makefile`?

markrtuttle replied:
Good catch.  There is now a makefile to automate the five or six invocations of doxygen.  The current build command is "make".

NlightNFotis commented on doc/doxygen-root/README.md:
```diff
@@ -0,0 +1,60 @@
+
+## Contribute documentation
+
+Coming soon...
```
There appears to be some documentation on this at `contributing.md`.

markrtuttle replied:
Added a link.  Thanks.

NlightNFotis commented on doc/doxygen-root/README.md:
```diff
@@ -0,0 +1,60 @@
+# CBMC documentation
```
Should this readme file be lifted to the top-level of the documentation folder, or does it conflict with the compilation of the documentation?

markrtuttle replied:
I could go either way.  My intention is that this documents the process of building the documentation, and so doesn't belong in the documentation itself.  I'm leaving it out for now.

NlightNFotis reviewed (approved):
Some minor comments that don't diminish the ability of this to move forward.

Thank you for your efforts to bring the documentation up to date.

feliperodri commented on doc/doxygen-root/developer_guide.md:
```diff
@@ -0,0 +1,14 @@
+* \ref cprover_documentation
+  * [CProver Architecture Decision Records](adr/index.html)
+  * [CProver APIs](api/index.html)
+  * [CProver assets](assets/index.html)
```
nitpick: `s/assets/Assets/g`

markrtuttle replied:
Done.

Commit:
Convert doc/ADR and doc/API to doxygen page/subpage syntax

Commit:
Correct and elaborate the documentation-build documentation

Commit:
Fix typo in developer guide
